### ISSUE TITLE
Writing directly to cache does not synchronously update `useQuery` return values

### ISSUE THREAD

**Issue by srmagura** [apollographql/apollo-client#9761]
**Intended outcome:**  When I write directly to my `InMemoryCache`, I want the updated data to be used the next time my React component renders.

**Actual outcome:** When I write directly to my `InMemoryCache`, there is a short time period in which my React component uses old data. 

This can cause bugs if you have code that assumes that writing to the cache **immediately** updates the relevant `useQuery` instances. Here's a real world example: imagine the user has just joined an organization in a multi-tenant application. The code might do the following:
1. Write to the cache to add a new organization to `currentUser.organizations`.
2. Use React Router's `navigate` function to go to the new organization's URL.
3. (Now the code explodes because the URL specifies an organization which is not contained in `currentUser.organizations`.)

**How to reproduce the issue:**
1. [View my CodeSandbox](https://6qm19o.csb.app/).
2. Open the console in the right pane.
4. Click "Add 1 to rates".

You should get output similar to the following. Notice how "Direct read" and "useQuery" are out of sync in the first rerender.

```text
Direct read: 4.673000143247005
useQuery: 3.6730001432470056 
------------------- 
Direct read: 4.673000143247005
useQuery: 4.673000143247005
------------------- 
```

**Versions**
![image](https://user-images.githubusercontent.com/801549/170105209-949b8cf2-0347-4a48-8485-8293a518bc65.png)

**Comment by srmagura** [comment]
Thanks for the great explanation. Two questions:

1. Is the Apollo code behaving as intended?
2. Is the difference between `cache.writeQuery` and `client.writeQuery` documented anywhere in your official docs? I do not see anywhere where this is mentioned.

Let me know if I can help by submitting a documentation PR.

**Comment by srmagura** [comment]
Our real code is using `cache.modify`, and there does not seem to be an equivalent method on `client`. Here's a workaround:

```ts
client.cache.modify(...);

(client as any).queryManager.broadcastQueries();
```

The `as any` is necessary because `ApolloClient.queryManager` is a private property. But it's only private from TypeScript's point of view — the property is still accessible at runtime.

**Comment by benjamn** [comment]
Yes, that's a reasonable workaround for other cache methods that don't have `client.*` versions. TypeScript may also allow you to call `client["queryManager"].broadcastQueries()`, for a little more type safety, depending on your TS version/config.

Are you looking for a quick way to restore previous behavior, or do you foresee wanting cache updates to be applied synchronously in the longer term?

**Comment by srmagura** [comment]
> Are you just looking for a quick way to restore previous behavior, or do you foresee wanting cache updates to be applied synchronously in the longer term?

I can foresee wanting cache updates to be applied synchronously in the longer term. Synchronous updates are useful for situations like the one I described in the original post. There is also just less potential for bugs when things are synchronous. Though, I understand that batching updates can be more efficient.

I think the bigger issue is that a method like `cache.modify` appears to be synchronous since it does not return a `Promise`, but its actual behavior is asynchronous. `modify` updates the cache immediately, but the method is still asynchronous when viewed in the larger context of a React application. 

This discrepancy between the method's expected and actual behavior could be remedied in two ways: 

A) Make `cache.modify` update all observers synchronously, OR
B) Make `cache.modify` return a `Promise` that resolves once all observers have been updated.

Since Apollo Client is (more or less) a global state manager, it could be useful to look at what other global state management libraries do. For example, I am pretty sure that Redux + react-redux does not do any batching unless you explicitly use the `batch` function that they export.

**Comment by ryanwilsonperkin** [comment]
I'm also interested in this from the perspective of `fetchMore`. It's possible in testing to get a reference to the invocation of a `fetchMore` call which resolves as a promise, but the actual work of updating the cache is deferred to a separate asynchronous process for which there's no way to get a handle.

As a result, this means that awaiting `fetchMore` within an `act()` scope, doesn't guarantee that the component will have enqueued a new render by the time the fetchMore promise resolves.
